OpenRA.Game/FieldLoader.cs

```diff
@@ -52,19 +58,33 @@ public static void Load(object self, MiniYaml my)
 			{
 				object val;
 
+				if (md == null)
+					md = my.ToDictionary();
 				if (fli.Loader != null)
-					val = fli.Loader(my);
+				{
+					if (!fli.Attribute.Required || md.ContainsKey(fli.YamlName))
+						val = fli.Loader(my);
+					else
+					{
+						missing.Add(fli);
+						continue;
+					}
+				}
 				else
 				{
-					if (md == null)
-						md = my.ToDictionary();
-
 					if (!TryGetValueFromYaml(fli.YamlName, fli.Field, md, out val))
+					{
+						if (fli.Attribute.Required)
+							missing.Add(fli);
 						continue;
+					}
 				}
 
 				fli.Field.SetValue(self, val);
 			}
+
+			if (missing.Any())
+				throw new MissingFieldsException(missing.First().YamlName + string.Concat(missing.Skip(1).Select(r => ", " + r.YamlName)));
```

Comment:
If you added a `string[]` property for the missing fields then you won't need to detect on `,` when handling it (which is a bit gross).